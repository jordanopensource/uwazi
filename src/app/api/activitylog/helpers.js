import { availableLanguages } from 'shared/languagesList';
import { typeParsers } from 'api/activitylog/migrationsParser';
import templates from 'api/templates/templates';
import entities from 'api/entities/entities';
import users from 'api/users/users';
import userGroups from 'api/usergroups/userGroups';
import { files } from 'api/files';
import { PermissionType } from 'shared/types/permissionSchema';
import { Suggestions } from 'api/suggestions/suggestions';
import { Extractors } from 'api/services/informationextraction/ixextractors';

const formatLanguage = langKey => {
  const lang = availableLanguages.find(({ key }) => key === langKey);
  return lang ? `${lang.label} (${lang.key})` : langKey;
};

const formatDataLanguage = data => formatLanguage(data.key);

const translationsName = data => {
  const [context] = data.contexts;
  return data.contexts.length === 1
    ? `in ${context.label} (${context.id})`
    : 'in multiple contexts';
};

const nameFunc = data => `${data.label} (${data.key})`;

const migrationLog = log => {
  const data = JSON.parse(log.body);
  return typeParsers[data.type] ? typeParsers[data.type](data) : { action: 'RAW' };
};

const templateName = data =>
  data.templateData ? `${data.templateData.name} (${data._id})` : data._id;

const loadEntityFromPublicForm = async data => {
  const entity = JSON.parse(data.entity);
  const templateData = await templates.getById(entity.template);
  return { ...data, templateData, title: entity.title };
};

const loadTemplate = async data => {
  let templateId = data.template || data._id;

  if (data.entity) {
    const entity = JSON.parse(data.entity);
    templateId = entity.template;
  }

  const templateData = await templates.getById(templateId);
  return { ...data, templateData };
};

const loadEntity = async data => {
  const _id = data.entityId || data._id;
  const sharedId = data.sharedId || data.entity;
  const query = { ...(_id && { _id }), ...(sharedId && { sharedId }) };
  const [entity] = await entities.getUnrestricted(query);
  return { ...data, entity, title: entity ? entity.title : undefined };
};

const loadFile = async data => {
  const [file] = await files.get({ _id: data._id });
  return { ...data, file, title: file ? file.originalname : `id: ${data._id}` };
};

const extraTemplate = data =>
  `of type ${
    data.templateData
      ? data.templateData.name
      : `(${data.template ? data.template.toString() : 'unassigned'})`
  }`;

const extraAttachmentLanguage = data =>
  data.entity
    ? `of entity '${data.entity.title}' (${data.entity.sharedId}) ${formatLanguage(
        data.entity.language
      )} version`
    : null;

const updatedFile = data => {
  let name;
  if (data.toc) {
    name = 'ToC, ';
  } else {
    name = '';
  }
  return `${name}${data.title}`;
};

const groupMembers = data => {
  const members = data.members.map(member => member.username).join(', ');
  return members.length > 0 ? `with members: ${members}` : 'with no members';
};

const loadPermissionsData = async data => {
  const updateEntities = await entities.getUnrestricted(
    { sharedId: { $in: data.ids } },
    { title: 1 }
  );
  const permissionsIds = data.permissions
    .filter(p => p.type !== PermissionType.PUBLIC)
    .map(pu => pu.refId);
  const allowedUsers = await users.get({ _id: { $in: permissionsIds } }, { username: 1 });
  const allowedGroups = await userGroups.get(
    { _id: { $in: permissionsIds } },
    { name: 1, members: 1 }
  );
  const publicPermission = !!data.permissions.find(p => p.type === PermissionType.PUBLIC);

  return {
    ...data,
    entities: updateEntities,
    users: allowedUsers,
    userGroups: allowedGroups,
    public: publicPermission,
  };
};

const entitiesNames = data => data.entities.map(e => e.title).join(', ');

function getNameOfAllowedPeople(source, field) {
  return p => {
    const people = source.find(u => u._id.toString() === p.refId);
    return `${people && people[field] ? people[field] : p.refId} - ${p.level}`;
  };
}

const loadAllowedUsersAndGroups = data => {
  const usersPermissions = data.permissions.filter(p => p.type === PermissionType.USER);
  const groupsPermissions = data.permissions.filter(p => p.type === PermissionType.GROUP);
  const grantedUsers = usersPermissions
    .map(getNameOfAllowedPeople(data.users, 'username'))
    .join(', ');
  const grantedNames = groupsPermissions
    .map(getNameOfAllowedPeople(data.userGroups, 'name'))
    .join(', ');

  return ` with permissions for${grantedUsers.length ? ` USERS: ${grantedUsers};` : ''}${
    grantedNames.length ? ` GROUPS: ${grantedNames}` : ''
  }${data.public ? '; PUBLIC' : ''}`;
};

const loadSuggestionData = async data => {
  const suggestion = await Suggestions.getById(data.suggestion._id);
  const entity = await entities.getById(data.suggestion.entityId);
  const [extractor] = await Extractors.get({ _id: suggestion.extractorId });
  return { ...data, ...suggestion, title: entity?.title, extractorName: extractor.name };
};

const loadExtractorData = async data => {
  const [extractor] = await Extractors.get({ _id: data.extractorId });
  return { ...data, ...extractor };
};

const loadUser = async data => {
  let [user] = await users.get({ _id: data._id });
  user = user || { username: data._id.toString() };
  return { ...data, ...user };
};

export {
  translationsName,
  formatLanguage,
  formatDataLanguage,
  nameFunc,
  migrationLog,
  templateName,
  loadEntityFromPublicForm,
  loadTemplate,
  loadEntity,
  loadFile,
  extraTemplate,
  extraAttachmentLanguage,
  updatedFile,
  groupMembers,
  loadPermissionsData,
  entitiesNames,
  loadAllowedUsersAndGroups,
  loadSuggestionData,
  loadExtractorData,
  loadUser,
};
