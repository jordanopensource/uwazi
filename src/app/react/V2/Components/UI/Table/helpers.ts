import { UniqueIdentifier } from '@dnd-kit/core';
import { cloneDeep } from 'lodash';
import { TableRow, TableProps } from './Table';

const getRowIds = <T extends TableRow<T>>(data: TableProps<T>['data']) => {
  const identifiers: { id: UniqueIdentifier; parentId?: string }[] = [];

  data.forEach(element => {
    identifiers.push({ id: element.rowId });

    if (element.subRows?.length) {
      element.subRows.forEach(subElement => {
        identifiers.push({ id: subElement.rowId, parentId: element.rowId });
      });
    }

    if (element.subRows?.length === 0) {
      identifiers.push({ id: `${element.rowId}-dropzone`, parentId: element.rowId });
    }
  });

  return identifiers;
};

// eslint-disable-next-line max-statements
const dndSortHandler = <T extends TableRow<T>>({
  currentState,
  dataIds,
  activeId,
  overId,
  disableEditingGroups,
}: {
  currentState: TableProps<T>['data'];
  dataIds: { id: UniqueIdentifier; parentId?: string }[];
  activeId: string | number;
  overId: string | number;
  disableEditingGroups?: boolean;
}): TableProps<T>['data'] => {
  const state = cloneDeep(currentState);

  const { activeParent, overParent } = dataIds.reduce(
    (acc, { id, parentId }) => {
      if (id === activeId) acc.activeParent = parentId;
      if (id === overId) acc.overParent = parentId;
      return acc;
    },
    { activeParent: undefined, overParent: undefined } as {
      activeParent?: string;
      overParent?: string;
    }
  );

  if (disableEditingGroups && activeParent !== overParent) {
    return currentState;
  }

  const { active: activePosition = 0, dropped: droppedPosition = 0 } = state.reduce(
    // eslint-disable-next-line max-statements
    ({ active, dropped }, item, index) => {
      let droppedIndex = dropped;
      let activeIndex = active;

      if (active === undefined) {
        if (item.rowId === activeParent) {
          activeIndex = item.subRows?.findIndex(element => element.rowId === activeId) || 0;
        } else if (item.rowId === activeId) {
          activeIndex = index;
        }
      }

      if (dropped === undefined) {
        if (item.rowId === overParent) {
          droppedIndex = item.subRows?.findIndex(element => element.rowId === overId) || 0;
        } else if (item.rowId === overId) {
          droppedIndex = index;
        }
      }

      return {
        active: activeIndex,
        dropped: droppedIndex,
      };
    },
    {
      active: undefined,
      dropped: undefined,
    } as {
      active?: number;
      dropped?: number;
    }
  );

  const activeElement = activeParent
    ? state.find(item => item.rowId === activeParent)?.subRows?.splice(activePosition, 1)[0]
    : state.splice(activePosition || 0, 1)[0];

  if (!activeElement || (Object.hasOwn(activeElement, 'subRows') && overParent)) {
    return currentState;
  }

  if (overParent) {
    state
      .find(item => item.rowId === overParent)
      ?.subRows?.splice(droppedPosition, 0, activeElement);
  } else {
    state.splice(droppedPosition, 0, activeElement);
  }

  return state;
};

export { getRowIds, dndSortHandler };
