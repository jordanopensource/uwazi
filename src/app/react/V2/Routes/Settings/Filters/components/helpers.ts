import { ClientSettingsFilterSchema } from 'app/apiResponseTypes';
import { ClientTemplateSchema } from 'app/istore';

type Filter = ClientSettingsFilterSchema & {
  rowId: string;
  subRows?: {
    rowId: string;
    _id?: string;
    id?: string;
    name?: string;
  }[];
};

type LoaderData = {
  filters: Filter[] | undefined;
  templates: ClientTemplateSchema[];
};

const filterAvailableTemplates = (templates: ClientTemplateSchema[], filters?: Filter[]) => {
  const usedTemplatesIds: string[] = [];

  filters?.forEach(filter => {
    if (filter.subRows) {
      filter.subRows.forEach(item => {
        usedTemplatesIds.push(item.id!);
      });
    }

    if (filter.id) {
      usedTemplatesIds.push(filter.id);
    }
  });

  return templates.filter(template => !usedTemplatesIds.includes(template._id));
};

const createNewFilters = (
  selectedTemplatesIds: string[],
  templates?: ClientTemplateSchema[]
): Filter[] => {
  const newFilters = selectedTemplatesIds.map(templateId => {
    const template = templates?.find(templ => templ._id === templateId);
    return { id: templateId, name: template?.name, rowId: templateId };
  });

  return newFilters;
};

const updateFilters = (newFilter: Filter, filters?: Filter[]) => {
  let isNewFilter = true;

  const updatedFilters = filters?.map(filter => {
    if (filter.id === newFilter.id) {
      isNewFilter = false;
      return newFilter;
    }
    return filter;
  });

  if (isNewFilter) {
    return [...(updatedFilters || []), newFilter];
  }

  return updatedFilters;
};

const deleteFilters = (originalFilters?: Filter[], filtersToRemove?: (string | undefined)[]) => {
  if (!filtersToRemove) {
    return originalFilters;
  }

  const updatedFilters: Filter[] = [];

  originalFilters?.forEach(filter => {
    const updatedFilter = { ...filter };
    if (!filtersToRemove.includes(filter.rowId)) {
      if (filter.subRows) {
        const subRows = filter.subRows.filter(item => !filtersToRemove.includes(item.rowId));
        updatedFilter.subRows = subRows;
      }
      updatedFilters.push(updatedFilter);
    }
  });

  return updatedFilters;
};

const sanitizeFilters = (filters?: Filter[]) => {
  const sanitizedFilters: ClientSettingsFilterSchema[] = [];

  filters?.forEach(filter => {
    const { rowId, subRows, ...sanitizedFilter } = { ...filter };

    if (subRows && subRows.length === 0) {
      return;
    }

    if (subRows) {
      sanitizedFilter.items = subRows.map(
        ({ rowId: itemRowId, _id, ...sanitizedItem }) => sanitizedItem
      );
    }

    sanitizedFilters.push(sanitizedFilter);
  });

  return sanitizedFilters;
};

const formatFilters = (filters: ClientSettingsFilterSchema[]): Filter[] =>
  filters?.map(filter => {
    const tableFilter: Filter = {
      ...filter,
      rowId: filter._id!,
    };
    if (filter.items) {
      const subRows = filter.items.map(item => ({ ...item, rowId: item.id! }));
      tableFilter.subRows = subRows;
    }
    return tableFilter;
  });

export type { LoaderData, Filter };
export {
  filterAvailableTemplates,
  createNewFilters,
  updateFilters,
  deleteFilters,
  sanitizeFilters,
  formatFilters,
};
