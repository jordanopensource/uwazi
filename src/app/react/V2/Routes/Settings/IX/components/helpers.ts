/* eslint-disable max-statements */
import { ClientEntitySchema } from 'app/istore';
import { PropertySchema } from 'shared/types/commonTypes';
import {
  SuggestionValue,
  TableSuggestion,
  SingleValueSuggestion,
  MultiValueSuggestion,
} from '../types';

const generateChildrenRows = (_suggestion: MultiValueSuggestion) => {
  const suggestion: MultiValueSuggestion = { ..._suggestion, isChild: false };

  const currentValues = [
    ...(Array.isArray(suggestion.currentValue) ? suggestion.currentValue : []),
  ];

  const suggestedValues = [
    ...(Array.isArray(suggestion.suggestedValue) ? suggestion.suggestedValue : []),
  ];

  suggestion.subRows = [];

  const { subRows, ...suggestionWithoutChildren } = suggestion;
  suggestedValues.forEach(suggestedValue => {
    const valuePresent = currentValues.find(v => v === suggestedValue);
    if (valuePresent) {
      currentValues.splice(currentValues.indexOf(valuePresent), 1);
    }

    suggestion.subRows?.push({
      ...suggestionWithoutChildren,
      suggestedValue,
      currentValue: valuePresent || '',
      propertyName: suggestion.propertyName,
      disableRowSelection: true,
      isChild: true,
      entityTitle: '',
      rowId: `${suggestion.rowId}-${suggestedValue}`,
    });
  });

  currentValues.forEach(currentValue => {
    suggestion.subRows?.push({
      ...suggestionWithoutChildren,
      suggestedValue: '',
      currentValue,
      disableRowSelection: true,
      isChild: true,
      entityTitle: '',
      rowId: `${suggestion.rowId}-${currentValue}`,
    });
  });

  return suggestion;
};

const _replaceSuggestion = (suggestion: TableSuggestion, currentSuggestions: TableSuggestion[]) =>
  currentSuggestions.map(currentSuggestion => {
    if (currentSuggestion._id === suggestion._id) {
      return suggestion;
    }
    return currentSuggestion;
  });

const _updateTitleInSuggestion = (
  currentSuggestions: TableSuggestion[],
  _suggestionToUpdate: TableSuggestion,
  updatedEntity: ClientEntitySchema
) => {
  const suggestionToUpdate = { ..._suggestionToUpdate };
  const newTitle = updatedEntity.title;
  suggestionToUpdate.currentValue = newTitle;
  suggestionToUpdate.entityTitle = newTitle!;
  suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === newTitle;

  return _replaceSuggestion(suggestionToUpdate, currentSuggestions);
};

// eslint-disable-next-line max-statements
const updateSuggestionsByEntity = (
  currentSuggestions: TableSuggestion[],
  updatedEntity?: ClientEntitySchema,
  property?: PropertySchema
): TableSuggestion[] => {
  if (!updatedEntity) {
    return currentSuggestions;
  }

  let suggestionToUpdate = currentSuggestions.find(
    currentSuggestion => currentSuggestion.entityId === updatedEntity._id
  );

  const propertyToUpdate = suggestionToUpdate?.propertyName;

  if (!suggestionToUpdate || !propertyToUpdate) {
    return currentSuggestions;
  }

  if (propertyToUpdate === 'title' && updatedEntity.title) {
    return _updateTitleInSuggestion(currentSuggestions, suggestionToUpdate, updatedEntity);
  }

  if (!updatedEntity.metadata) {
    return currentSuggestions;
  }

  if (updatedEntity.metadata[propertyToUpdate]?.length) {
    const newValue = (
      property?.type === 'multiselect' || property?.type === 'relationship'
        ? updatedEntity.metadata[propertyToUpdate]?.map(v => v.value)
        : updatedEntity.metadata[propertyToUpdate]![0].value
    ) as SuggestionValue;

    suggestionToUpdate.currentValue = newValue;
    suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === newValue;
  }

  if (!updatedEntity.metadata[propertyToUpdate]?.length) {
    suggestionToUpdate.currentValue = '';
    suggestionToUpdate.state.match = suggestionToUpdate.suggestedValue === '';
  }

  if (property?.type === 'multiselect' || property?.type === 'relationship') {
    suggestionToUpdate = generateChildrenRows(suggestionToUpdate as MultiValueSuggestion);
  }

  return _replaceSuggestion(suggestionToUpdate, currentSuggestions);
};

const updateMultiValueSuggestions = (
  _parentSuggestion: MultiValueSuggestion,
  acceptedSuggestion: SingleValueSuggestion
): MultiValueSuggestion => {
  let parentSuggestion = { ..._parentSuggestion };

  const shouldAddValue = acceptedSuggestion.suggestedValue !== '';
  const value = shouldAddValue
    ? acceptedSuggestion.suggestedValue || ''
    : acceptedSuggestion.currentValue || '';

  parentSuggestion.currentValue = (_parentSuggestion.currentValue as SuggestionValue[]) || [];

  parentSuggestion.currentValue = shouldAddValue
    ? parentSuggestion.currentValue.concat(value)
    : parentSuggestion.currentValue.filter(v => v !== value);

  parentSuggestion.state.match = parentSuggestion.currentValue.every(v =>
    parentSuggestion.suggestedValue.includes(v)
  );

  parentSuggestion = generateChildrenRows(parentSuggestion);
  return parentSuggestion;
};

const updateSuggestions = (
  currentSuggestions: TableSuggestion[],
  suggestionsToAccept: TableSuggestion[]
): TableSuggestion[] => {
  if (!suggestionsToAccept.length) {
    return currentSuggestions;
  }

  const acceptedSuggestions = suggestionsToAccept.map(acceptedSuggestion => {
    let suggestion = (
      acceptedSuggestion.isChild
        ? { ...currentSuggestions.find(s => s._id === acceptedSuggestion._id) }
        : { ...acceptedSuggestion }
    ) as TableSuggestion;

    if (acceptedSuggestion.isChild) {
      suggestion = updateMultiValueSuggestions(
        suggestion as MultiValueSuggestion,
        acceptedSuggestion as SingleValueSuggestion
      );
    } else {
      suggestion.state.match = true;
      suggestion.currentValue = acceptedSuggestion.suggestedValue;
    }

    if ('subRows' in suggestion && suggestion.subRows?.length) {
      suggestion = generateChildrenRows(suggestion as MultiValueSuggestion);
    }

    if (
      acceptedSuggestion.propertyName === 'title' &&
      typeof acceptedSuggestion.suggestedValue === 'string'
    ) {
      suggestion.entityTitle = acceptedSuggestion.suggestedValue;
    }
    return suggestion;
  });

  const merged = [
    ...currentSuggestions
      .concat(acceptedSuggestions)
      .reduce(
        (map, suggestion) =>
          map.set(suggestion._id, Object.assign(map.get(suggestion._id) || {}, suggestion)),
        new Map()
      )
      .values(),
  ];

  return merged;
};

export { updateSuggestions, updateSuggestionsByEntity, generateChildrenRows };
