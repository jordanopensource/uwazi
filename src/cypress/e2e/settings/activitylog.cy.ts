import 'cypress-axe';
import { clearCookiesAndLogin } from '../helpers';

describe('Activity log', () => {
  before(() => {
    const env = { DATABASE_NAME: 'uwazi_e2e', INDEX_NAME: 'uwazi_e2e' };
    cy.exec('yarn e2e-fixtures', { env });
    clearCookiesAndLogin('editor', 'editor');
  });

  it('should register entity creation in activity log', () => {
    cy.contains('a', 'Library').click();
    cy.contains('Create entity').click();
    cy.get('#metadataForm select').select('Reporte');
    cy.contains('Title').scrollIntoView();
    cy.get('textarea[name="library.sidepanel.metadata.title"]:not([disabled])').type('AL Report');
    cy.contains('aside.side-panel.metadata-sidepanel.is-active .btn-success', 'Save').click();
    cy.contains('Entity created');
  });

  it('should register entity edition in activity log', () => {
    cy.contains('div.item-document', 'AL Report').click();
    cy.contains('.metadata-sidepanel.is-active .edit-metadata', 'Edit').click();
    cy.get('#metadataForm select').eq(0).select('Reporte');
    cy.contains('Title').scrollIntoView();
    cy.clearAndType(
      'textarea[name="library.sidepanel.metadata.title"]:not([disabled])',
      'Report AL'
    );
    cy.contains('.metadata-sidepanel.is-active .btn-success', 'Save').click();
    cy.contains('Entity updated');
  });

  it('should register entity deletion in activity log', () => {
    cy.contains('div.item-document', 'Report AL', { timeout: 200 }).click();
    cy.contains('.metadata-sidepanel.is-active .btn-danger', 'Delete').click();
    cy.contains('.confirm-button', 'Accept').click();
    cy.contains('Entity deleted');
  });

  it('should register account edition in activity log', () => {
    cy.get('.only-desktop a[aria-label="Settings"]').click();
    cy.contains('span', 'Account').click();
    cy.contains('button', 'Enable').click();
    cy.contains('button', 'Cancel').click();
    cy.clearAndType('input[name=email]', 'editor@uwazi.com', { delay: 0 });
    cy.contains('button', 'Update').click();
    cy.get('[data-testid="modal"]').within(() => {
      cy.get('input').type('editor');
      cy.contains('button', 'Accept').click();
    });
  });

  it('should pass accessibility check', () => {
    clearCookiesAndLogin();
    cy.get('.only-desktop a[aria-label="Settings"]').click();
    cy.contains('span', 'Activity log').click();
    cy.contains('editor');
    cy.injectAxe();
    cy.checkA11y();
  });

  const checkCells = (row: number, column: number, element: string) => {
    cy.get(`tr:nth-child(${row}) > td:nth-child(${column}) > ${element}`).toMatchSnapshot();
  };

  it('should list the last activity log entries', () => {
    cy.get('tr').should('have.length.at.least', 8);
    checkCells(1, 1, 'span');
    checkCells(1, 2, 'span');
    checkCells(1, 3, ' div > div:nth-child(1)');
    checkCells(2, 1, 'span');
    checkCells(2, 2, 'span');
    checkCells(2, 3, ' div > div:nth-child(1)');
  });

  const applyFilters = () => {
    cy.intercept('GET', 'api/activityLog*').as('getActivityLog');
    cy.contains('button', 'Apply').click();
    cy.wait('@getActivityLog');
    cy.contains('button', 'Apply').should('not.exist');
  };

  it('should filter by user', () => {
    cy.contains('button', 'Filters').click();
    cy.contains('Apply');
    cy.get('input[name=username]').type('editor', { delay: 0 });
    applyFilters();
    cy.contains('Updated user');
    cy.get('tr').should('have.length.at.most', 6);
  });

  it('should show a tooltip with the detail of an activity entry', () => {
    cy.contains('button', 'Filters').click();

    applyFilters();
    cy.contains('RAW').should('not.exist');
    cy.get('tr:nth-child(1) td:nth-child(3)').within(() => {
      cy.contains('Updated user').invoke('show').trigger('mouseenter');
    });
    cy.get('tr:nth-child(1) td:nth-child(3)').within(() => {
      cy.contains('Updated user').invoke('show').trigger('mouseenter');
    });
    cy.contains('Body');
    cy.get('tr:nth-child(1) td:nth-child(3) [data-testid=flowbite-tooltip]')
      .should('be.visible')
      .invoke('text')
      .should('match', /Query{}Body{.+"username":"editor".+"email":"editor@uwazi.com"}/);
    cy.get('tr:nth-child(1) td:nth-child(3)').within(() => {
      cy.contains('Updated user').trigger('mouseleave');
    });
  });

  it('should update the list when filters are cleaned', () => {
    cy.contains('button', 'Filters').click();
    cy.contains('button', 'Clear all').click();
    applyFilters();
    cy.get('tr').should('have.length.at.least', 8);
  });

  it('should open the detail of an entry', () => {
    cy.contains('Updated user');
    cy.get('tr:nth-child(1) td:nth-child(5)').contains('View').click({ force: true });
    cy.contains('Query', { timeout: 200 });
    cy.get('aside.ease-in').toMatchSnapshot();
    cy.get('[data-testid=close-sidepanel]').click();
  });

  const selectMethod = (option: number) => {
    cy.getByTestId('multiselect').within(() => {
      cy.get('button').eq(0).click();
      cy.get('ul li')
        .eq(option)
        .within(() => {
          cy.get('input').eq(0).click();
        });
    });
  };

  it('should filter by method', () => {
    cy.contains('button', 'Filters').click();
    selectMethod(2);
    applyFilters();
    cy.contains('editor');
    cy.get('tr').should('have.length.at.most', 3);
  });

  it('should filter by dates', () => {
    cy.contains('button', 'Filters').click();
    cy.contains('button', 'Clear all').click();
    applyFilters();
    cy.contains('button', 'Filters').click();
    cy.get('#from').type('2024-04-25', { delay: 0 });
    cy.get('#to').clear();
    cy.get('#to').type('2024-05-09', { delay: 0 });
    cy.get('#to').realPress('Tab');
    applyFilters();
    cy.get('tr').should('have.length.at.most', 5);
  });

  // eslint-disable-next-line max-statements
  it('should do a composed filter with search', () => {
    cy.contains('button', 'Filters').click();
    cy.contains('button', 'Clear all').click();
    cy.clearAndType('input[name=username]', 'editor', { delay: 0 });
    cy.clearAndType('input[name=search]', 'Deleted entity', { delay: 0 });
    selectMethod(2);
    cy.get('#from').type('2024-05-28', { delay: 0 });
    cy.get('#to').click();
    cy.get('.datepicker:not(.hidden) .datepicker-controls .today-btn').eq(1).click();
    applyFilters();
    cy.get('tr').should('have.length', 2);
    cy.contains('Deleted entity');
  });

  // eslint-disable-next-line max-statements
  it('should do a composed filter with today', () => {
    cy.contains('button', 'Filters').click();
    cy.contains('button', 'Clear all').click();
    cy.clearAndType('input[name=username]', 'editor', { delay: 0 });
    selectMethod(0);
    cy.get('#from').type('2024-05-28', { delay: 0 });
    cy.get('#to').click();
    cy.get('.datepicker:not(.hidden) .datepicker-controls .today-btn').eq(1).click();
    applyFilters();
    cy.get('tr').should('have.length', 3);
    cy.get('table > tbody > tr:nth-child(2) > td:nth-child(3)').contains('Created entity:');
    cy.get('table > tbody > tr:nth-child(2) > td:nth-child(3)').contains('AL Report');
  });
});
