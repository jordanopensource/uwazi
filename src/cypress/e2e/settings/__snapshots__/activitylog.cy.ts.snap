exports[`Activity log > should list the last activity log entries #0`] = `
<span
  class="  bg-blue-100 text-blue-800 px-2.5 py-1 rounded-md text-xs "
  data-testid="pill-comp"
  ><span class="translation">UPDATE</span></span
>
`;

exports[`Activity log > should list the last activity log entries #1`] = `
<span class="text-primary-700">editor</span>
`;

exports[`Activity log > should list the last activity log entries #2`] = `
<div class="w-fit" data-testid="flowbite-tooltip-target">
  <div class="space-x-2">
    <span class="translation font-semibold">Updated user</span>:<span
      class="translation"
      >editor (58ada34d299e82674854510e)</span
    >
  </div>
</div>
`;

exports[`Activity log > should list the last activity log entries #3`] = `
<span
  class="  bg-green-100 text-green-800 px-2.5 py-1 rounded-md text-xs "
  data-testid="pill-comp"
  ><span class="translation">CREATE</span></span
>
`;

exports[`Activity log > should list the last activity log entries #4`] = `
<span class="text-primary-700">editor</span>
`;

exports[`Activity log > should list the last activity log entries #5`] = `
<div class="w-fit" data-testid="flowbite-tooltip-target">
  <div class="space-x-2">
    <span class="translation font-semibold"
      >Two-factor authentication secret</span
    >:
  </div>
</div>
`;

exports[`Activity log > should open the detail of an entry #0`] = `
<aside
  class="w-full h-full top-0 right-0 fixed bg-white border-l-2 transition duration-200 ease-in transform md:w-[500px] -translate-x-[-500px]"
>
  <div class="flex flex-col h-full overflow-y-auto">
    <div class="flex p-4 mb-2 text-gray-500">
      <h1 class="text-base font-bold grow">
        <span class="translation uppercase">Activity</span>
      </h1>
      <button type="button" class="justify-end" data-testid="close-sidepanel">
        <span class="sr-only"><span class="translation">Close</span></span
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 20 20"
          fill="currentColor"
          aria-hidden="true"
          data-slot="icon"
          class="w-6"
        >
          <path
            d="M6.28 5.22a.75.75 0 0 0-1.06 1.06L8.94 10l-3.72 3.72a.75.75 0 1 0 1.06 1.06L10 11.06l3.72 3.72a.75.75 0 1 0 1.06-1.06L11.06 10l3.72-3.72a.75.75 0 0 0-1.06-1.06L10 8.94 6.28 5.22Z"
          ></path>
        </svg>
      </button>
    </div>
    <div class="flex-grow p-4 ">
      <div class="flex flex-col">
        <div
          class="flex flex-row justify-between w-full p-5 my-2 rounded-lg bg-gray-50"
        >
          <div class="flex-col">
            <div>
              <span class="translation font-semibold text-primary-700"
                >Updated user</span
              >:
            </div>
            <div>
              <span class="translation">editor (58ada34d299e82674854510e)</span>
            </div>
          </div>
          <span
            class="h-6  bg-blue-100 text-blue-800 px-2.5 py-1 rounded-md text-xs "
            data-testid="pill-comp"
            ><span class="translation">UPDATE</span></span
          >
        </div>
      </div>
      <div
        class="flex flex-row justify-between w-full p-5 my-2 border rounded-lg border-gray-50 bg-gray-30"
      >
        <div class="flex flex-col">
          <span class="translation font-semibold">User</span
          ><span class="text-primary-700">editor</span>
        </div>
      </div>
      <div class="flex flex-col flex-grow p-3 rounded-lg max-h-svh h-3/4">
        <span class="translation m-2 font-semibold">Query</span
        ><span
          class="block p-5 mb-4 text-gray-900 border border-gray-300 rounded-lg bg-gray-50"
          >{}</span
        ><span class="translation m-2 font-semibold">Body</span
        ><span
          class="block w-full p-3 overflow-auto text-gray-900 border border-gray-300 rounded-lg bg-gray-50 "
          >{"_id":"58ada34d299e82674854510e","username":"editor","__v":0,"role":"editor","groups":[{"_id":"5fda2b675917fe58048a88f4","name":"Activistas"},{"_id":"5fda2b6f5917fe58048a88f6","name":"Asesores
          legales"}],"email":"editor@uwazi.com"}</span
        >
      </div>
    </div>
  </div>
</aside>
`;

exports[`Activity log > should open the detail of an entry #1`] = `
<aside
  class="w-full h-full top-0 right-0 fixed bg-white border-l-2 transition duration-200 ease-in transform md:w-[500px] translate-x-0"
>
  <div class="flex flex-col h-full overflow-y-auto">
    <div class="flex p-4 mb-2 text-gray-500">
      <h1 class="text-base font-bold grow">
        <span class="translation uppercase">Activity</span>
      </h1>
      <button type="button" class="justify-end" data-testid="close-sidepanel">
        <span class="sr-only"><span class="translation">Close</span></span
        ><svg
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 20 20"
          fill="currentColor"
          aria-hidden="true"
          data-slot="icon"
          class="w-6"
        >
          <path
            d="M6.28 5.22a.75.75 0 0 0-1.06 1.06L8.94 10l-3.72 3.72a.75.75 0 1 0 1.06 1.06L10 11.06l3.72 3.72a.75.75 0 1 0 1.06-1.06L11.06 10l3.72-3.72a.75.75 0 0 0-1.06-1.06L10 8.94 6.28 5.22Z"
          ></path>
        </svg>
      </button>
    </div>
    <div class="flex-grow p-4 ">
      <div class="flex flex-col">
        <div
          class="flex flex-row justify-between w-full p-5 my-2 rounded-lg bg-gray-50"
        >
          <div class="flex-col">
            <div>
              <span class="translation font-semibold text-primary-700"
                >Updated user</span
              >:
            </div>
            <div>
              <span class="translation">editor (58ada34d299e82674854510e)</span>
            </div>
          </div>
          <span
            class="h-6  bg-blue-100 text-blue-800 px-2.5 py-1 rounded-md text-xs "
            data-testid="pill-comp"
            ><span class="translation">UPDATE</span></span
          >
        </div>
      </div>
      <div
        class="flex flex-row justify-between w-full p-5 my-2 border rounded-lg border-gray-50 bg-gray-30"
      >
        <div class="flex flex-col">
          <span class="translation font-semibold">User</span
          ><span class="text-primary-700">editor</span>
        </div>
      </div>
      <div class="flex flex-col flex-grow p-3 rounded-lg max-h-svh h-3/4">
        <span class="translation m-2 font-semibold">Query</span
        ><span
          class="block p-5 mb-4 text-gray-900 border border-gray-300 rounded-lg bg-gray-50"
          >{}</span
        ><span class="translation m-2 font-semibold">Body</span
        ><span
          class="block w-full p-3 overflow-auto text-gray-900 border border-gray-300 rounded-lg bg-gray-50 "
          >{"_id":"58ada34d299e82674854510e","username":"editor","__v":0,"role":"editor","groups":[{"_id":"5fda2b675917fe58048a88f4","name":"Activistas"},{"_id":"5fda2b6f5917fe58048a88f6","name":"Asesores
          legales"}],"email":"editor@uwazi.com"}</span
        >
      </div>
    </div>
  </div>
</aside>
`;
